src/toolbox/shamir_secrets.rs

```diff
@@ -7,14 +8,82 @@ pub struct SecretShare {
     threshold: usize,
 }
 
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Share {
+    pub x: Scalar,              // this is the x in f(x) of our polynomial
+    pub y: Scalar,              // this is the f(x) of our polynomial
+}
+
 impl SecretShare {
+
+    pub fn share(secret: &Scalar, nr_of_shares: usize, threshold: usize) -> SecretShare {
+        let mut coefficients: Vec<Scalar> = Vec::new();
+        let my_secret = secret.clone();
+        coefficients.push(my_secret);  // the coefficient for x^0 is the secret
+        for _ in 1..(threshold) {
+
+            // TODO should this be an independent RNG, or the transcript RNG?
+            // If it's the transcript's, anyone with the transcript would be able to generate the same polynomials
+            //      I don't think we want that, because you could write code to do so and easily recover the secret from just a single share
+            coefficients.push(Scalar::random(&mut rand::thread_rng()));
+        }
+
+        // println!("Polynomial coefficients: {:?}", coefficients);
+
+        let mut shares: Vec<Scalar> = Vec::new();
+        for i in 0..nr_of_shares {
+            let x = Scalar::from((i+1) as u64);
+            let y = SecretShare::calc_polynomial(&coefficients, &x);
+            shares.push(y);
+        }
+
+        // println!("Generated shares: {:?}", shares);
+        // select random polynomial of degree *threshold*
+        // 0-coeff: secret
+        // output the shares, which are points evaluated at i=1,2,3,...,nr_of_shares
+        SecretShare {secret: my_secret, shares, threshold}
+    }
+
+    fn calc_polynomial(coefficients: &Vec<Scalar>, x: &Scalar) -> Scalar {
+        let mut ret = Scalar::zero();
+
+        for i in 0..coefficients.len() {
+            if let Ok(xp) = SecretShare::pow(&x, i) {
+                ret += xp * coefficients[i];
+            }
+        };
+        ret
+    }
+
+    fn pow(x: &Scalar, p: usize) -> Result<Scalar, &'static str> {
```

Comment:
Why not return just a Scalar, if you are not going to throw any error here...

Reply:
Good point!